src/rust/engine/src/core.rs

```diff
@@ -351,23 +353,86 @@ impl Failure {
       }
     }
   }
-}
 
-impl Failure {
+  pub fn join_tracebacks(self, new_failure: Self) -> Failure {
+    match (self, new_failure) {
+      (
+        Failure::Throw {
+          val: _old_val,
+          python_traceback: old_python_traceback,
+          engine_traceback: old_engine_traceback,
+        },
+        Failure::Throw {
+          val: new_val,
+          python_traceback: new_python_traceback,
+          engine_traceback: new_engine_traceback,
+        },
+      ) => {
+        let joined_python_tracebacks = [
+          old_python_traceback,
+          "\nWhile handling that, another exception was raised:\n".to_string(),
+          new_python_traceback,
+        ]
+        .join("\n");
+        let joined_engine_tracebacks = old_engine_traceback
+          .into_iter()
+          .chain(new_engine_traceback.into_iter())
+          .collect::<Vec<_>>();
+        Failure::Throw {
+          val: new_val,
+          python_traceback: joined_python_tracebacks,
+          engine_traceback: joined_engine_tracebacks,
+        }
+      }
+      _ => Failure::Invalidated,
+    }
+  }
+
+  pub fn as_py_err(&self) -> Result<&Value, Self> {
+    match &self {
```

Comment:
Does the current version of rust let you get rid of all of these `&`s and still compile?

Reply:
Yes! Fixed!